### PR TITLE
Control Plane Configuration Controller

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Requires #296 

Implement `ControlPlaneConfigurationController` and use it in the microcluster app.

This controller watches over the cluster configuration and applies any required updates to the control plane services of the local node.

### CODE REVIEW

**Review by bschimke95 — COMMENTED** [review]
First pass, great work

File: src/k8s/pkg/k8sd/app/hooks_start.go [hunk 1]
```diff
@@ -13,5 +17,12 @@ func (a *App) onStart(s *state.State) error {
 		go a.nodeConfigController.Run(s.Context)
 	}
 
+	// start control plane config controller
+	if a.controlPlaneConfigController != nil {
+		go a.controlPlaneConfigController.Run(s.Context, func(ctx context.Context) (types.ClusterConfig, error) {
+			return utils.GetClusterConfig(ctx, s)
+		})
```
Comment (addyess) [inline src/k8s/pkg/k8sd/app/hooks_start.go:24]: go routines -- i haven't learned this in my training yet.  Kinda like python `async` maybe? or more like an async thread?

File: src/k8s/pkg/k8sd/app/cluster_util.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func setupControlPlaneServices(snap snap.Snap, s *state.State, cfg types.Cluster
 	if err := setup.KubeScheduler(snap); err != nil {
 		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
 	}
-	if err := setup.KubeAPIServer(snap, cfg.Network.GetServiceCIDR(), s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.Datastore.GetType(), cfg.Datastore.GetExternalURL(), cfg.APIServer.GetAuthorizationMode()); err != nil {
+	if err := setup.KubeAPIServer(snap, cfg.Network.GetServiceCIDR(), s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.Datastore, cfg.APIServer.GetAuthorizationMode()); err != nil {
```
[the remark this answers is not in the document]
Comment (addyess) [reply src/k8s/pkg/k8sd/app/cluster_util.go:60]: Agree.  The context on this PR is missing where you're changing the datastore argument bits...
* https://github.com/canonical/k8s-snap/pull/296

**Review by addyess — COMMENTED** [review]
Drive by comments

File: src/k8s/pkg/k8sd/app/hooks_start.go (hunk 1 above)
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/hooks_start.go:24]: Go routines are green threads (user-space). further, they are not colored separately (e.g. no need to do `func foo()` vs `async func foo()`. Let's discuss offline and go into more details

File: src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+			},
+		}
+
+		select {
+		case triggerCh <- time.Now():
+		case <-time.After(channelSendTimeout):
+			g.Fail("Timed out while attempting to trigger controller reconcile loop")
+		}
+
+		// TODO: this should be changed to call g.Eventually()
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go:283]: nit: what is currently blocking this?

**Review by bschimke95 — APPROVED** [review]
LGTM, only a minor question

File: src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go (hunk 3 above)
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go:283]: it's just about spending proper effort to rewrite the tests accordingly.

Comment (neoaggelos) [reply src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go:283]: can be done at a later time, not pressing atm

**Pull request merged**